cwlVersion: v1.0
class: Workflow


requirements:
  - class: SubworkflowFeatureRequirement
  - class: StepInputExpressionRequirement
  - class: InlineJavascriptRequirement
  - class: MultipleInputFeatureRequirement


inputs:

  indices_folder:
    type: Directory
    doc: |
      Cell Ranger reference genome indices folder

  fastq_file_r1:
    type: File
    doc: |
      FASTQ file R1 (optionally compressed)

  fastq_file_r2:
    type: File
    doc: |
      FASTQ file R2 (optionally compressed)

  expect_cells:
    type: int?
    default: 3000
    doc: |
      Expected number of recovered cells

  force_expect_cells:
    type: boolean?
    default: false
    doc: |
      Force pipeline to use the expected number of recovered cell.
      The value provided in expect_cells will be sent to Cell Ranger
      Count as --force-cells. The latter will bypass the cell detection
      algorithm. Use this if the number of cells estimated by Cell Ranger
      is not consistent with the barcode rank plot.

  include_introns:
    type: boolean?
    default: false
    doc: |
      Count reads mapping to intronic regions. This may improve sensitivity
      for samples with a significant amount of pre-mRNA molecules, such as nuclei.

  threads:
    type: int?
    default: 4
    doc: |
      Number of threads for those steps that support multithreading

  memory_limit:
    type: int?
    default: 20
    doc: |
      Maximum memory used (GB). The same will be applied to virtual memory


outputs:

  web_summary_report:
    type: File
    outputSource: count_gene_expr/web_summary_report
    doc: |
      Run summary metrics and charts in HTML format

  metrics_summary_report:
    type: File
    outputSource: count_gene_expr/metrics_summary_report
    doc: |
      Run summary metrics in CSV format

  possorted_genome_bam_bai:
    type: File
    outputSource: count_gene_expr/possorted_genome_bam_bai
    doc: |
      Indexed reads aligned to the genome and transcriptome annotated
      with barcode information

  filtered_feature_bc_matrix_folder:
    type: File
    outputSource: compress_filtered_feature_bc_matrix_folder/compressed_folder
    doc: |
      Compressed folder with filtered feature-barcode matrices containing only
      cellular barcodes in MEX format

  filtered_feature_bc_matrix_h5:
    type: File
    outputSource: count_gene_expr/filtered_feature_bc_matrix_h5
    doc: |
      Filtered feature-barcode matrices containing only cellular barcodes
      in HDF5 format
  
  raw_feature_bc_matrices_folder:
    type: File
    outputSource: compress_raw_feature_bc_matrices_folder/compressed_folder
    doc: |
      Compressed folder with unfiltered feature-barcode matrices containing
      all barcodes in MEX format

  raw_feature_bc_matrices_h5:
    type: File
    outputSource: count_gene_expr/raw_feature_bc_matrices_h5
    doc: |
      Unfiltered feature-barcode matrices containing all barcodes in HDF5 format

  secondary_analysis_report_folder:
    type: File
    outputSource: compress_secondary_analysis_report_folder/compressed_folder
    doc: |
      Compressed folder with secondary analysis results including dimensionality
      reduction, cell clustering, and differential expression

  molecule_info_h5:
    type: File
    outputSource: count_gene_expr/molecule_info_h5
    doc: |
      Molecule-level information used by cellranger aggr to aggregate samples
      into larger datasets

  loupe_browser_track:
    type: File
    outputSource: count_gene_expr/loupe_browser_track
    doc: |
      Loupe Browser visualization and analysis file

  compressed_html_data_folder:
    type: File
    outputSource: compress_html_data_folder/compressed_folder
    doc: |
      Compressed folder with UCSC Cellbrowser formatted results

  count_gene_expr_stdout_log:
    type: File
    outputSource: count_gene_expr/stdout_log
    doc: |
      stdout log generated by cellranger count

  count_gene_expr_stderr_log:
    type: File
    outputSource: count_gene_expr/stderr_log
    doc: |
      stderr log generated by cellranger count


steps:

  extract_fastq_r1:
    doc: |
      Extracts bzip2-/gzip-/zip-compressed FASTQ file
    run: ../tools/extract-fastq.cwl
    in:
      compressed_file: fastq_file_r1
    out:
    - fastq_file

  extract_fastq_r2:
    doc: |
      Extracts bzip2-/gzip-/zip-compressed FASTQ file
    run: ../tools/extract-fastq.cwl
    in:
      compressed_file: fastq_file_r2
    out:
    - fastq_file

  count_gene_expr:
    doc: |
      Quantifies gene expression from a single-cell RNA-Seq library
    run: ../tools/cellranger-count.cwl
    in:
      fastq_file_r1: extract_fastq_r1/fastq_file
      fastq_file_r2: extract_fastq_r2/fastq_file
      indices_folder: indices_folder
      expect_cells:
        source: [expect_cells, force_expect_cells]
        valueFrom: $(self[1]?null:self[0])
      force_cells:
        source: [expect_cells, force_expect_cells]
        valueFrom: $(self[1]?self[0]:null)
      include_introns: include_introns
      threads: threads
      memory_limit: memory_limit
      virt_memory_limit: memory_limit
    out:
    - web_summary_report
    - metrics_summary_report
    - possorted_genome_bam_bai
    - filtered_feature_bc_matrix_folder
    - filtered_feature_bc_matrix_h5
    - raw_feature_bc_matrices_folder
    - raw_feature_bc_matrices_h5
    - secondary_analysis_report_folder
    - molecule_info_h5
    - loupe_browser_track
    - stdout_log
    - stderr_log

  cellbrowser_build:
    doc: |
      Exports clustering results from Cell Ranger Count Gene Expression or Cell
      Ranger Aggregate experiments into compatible with UCSC Cell Browser format
    run: ../tools/cellbrowser-build-cellranger.cwl
    in:
      secondary_analysis_report_folder: count_gene_expr/secondary_analysis_report_folder
      filtered_feature_bc_matrix_folder: count_gene_expr/filtered_feature_bc_matrix_folder
    out:
    - html_data

  compress_filtered_feature_bc_matrix_folder:
    doc: |
      Creates compressed TAR file from a folder
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: count_gene_expr/filtered_feature_bc_matrix_folder
    out:
    - compressed_folder

  compress_raw_feature_bc_matrices_folder:
    doc: |
      Creates compressed TAR file from a folder
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: count_gene_expr/raw_feature_bc_matrices_folder
    out:
    - compressed_folder

  compress_secondary_analysis_report_folder:
    doc: |
      Creates compressed TAR file from a folder
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: count_gene_expr/secondary_analysis_report_folder
    out:
    - compressed_folder

  compress_html_data_folder:
    doc: |
      Creates compressed TAR file from a folder
    run: ../tools/tar-compress.cwl
    in:
      folder_to_compress: cellbrowser_build/html_data
    out:
    - compressed_folder


$namespaces:
  s: http://schema.org/

$schemas:
- https://github.com/schemaorg/schemaorg/raw/main/data/releases/11.01/schemaorg-current-http.rdf

label: "Single-cell RNA-Seq Alignment"
s:name: "Single-cell RNA-Seq Alignment"
s:alternateName: "Runs Cell Ranger Count to quantify gene expression from a single-cell RNA-Seq library"

s:downloadUrl: https://raw.githubusercontent.com/Barski-lab/scRNA-Seq-Analysis/main/workflows/sc-rna-align-wf.cwl
s:codeRepository: https://github.com/Barski-lab/scRNA-Seq-Analysis
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898


doc: |
  Single-cell RNA-Seq Alignment

  Runs Cell Ranger Count to quantify gene expression
  from a single-cell RNA-Seq library.